File: cli-creator/examples/python-crud-cli/build_calls.py
"""
Pure functions that transform CLI arguments into API call parameters.
No I/O, no side effects -- easy to test without mocking.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class ApiCall:
    tool: str
    args: dict[str, Any]


# --- Search ---


def build_search_call(
    query: str,
    *,
    limit: int | None = None,
    filter_type: str | None = None,
) -> dict[str, Any]:
    args: dict[str, Any] = {
        "query": query,
        "page_size": limit or 20,
    }
    if filter_type:
        args["filter"] = {"property": "object", "value": filter_type}
    return {"tool": "search", "args": args}


# --- Fetch ---


def build_fetch_call(
    page_id: str,
    *,
    include_children: bool = False,
) -> dict[str, Any]:
    args: dict[str, Any] = {"page_id": page_id}
    if include_children:
        args["include_children"] = True
    return {"tool": "get_page", "args": args}


# --- Create ---


def build_create_call(
    *,
    parent: str,
    title: str,
    body: str | None = None,
) -> dict[str, Any]:
    args: dict[str, Any] = {
        "parent_id": parent,
        "properties": {"title": title},
    }
    if body:
        args["children"] = [
            {
                "type": "paragraph",
                "paragraph": {
                    "rich_text": [{"type": "text", "text": {"content": body}}],
                },
            }
        ]
    return {"tool": "create_page", "args": args}


# --- Update ---


def build_update_call(
    page_id: str,
    *,
    title: str | None = None,
    archived: bool | None = None,
) -> dict[str, Any]:
    args: dict[str, Any] = {"page_id": page_id}
    properties: dict[str, Any] = {}

    if title is not None:
        properties["title"] = title
    if properties:
        args["properties"] = properties
    if archived is not None:
        args["archived"] = archived

    return {"tool": "update_page", "args": args}


# --- Delete (Archive) ---


def build_delete_call(page_id: str) -> dict[str, Any]:
    return build_update_call(page_id, archived=True)


# --- Database Query ---


def build_query_call(
    database_id: str,
    *,
    filter: dict[str, Any] | None = None,
    sort_property: str | None = None,
    sort_direction: str = "ascending",
    limit: int | None = None,
) -> dict[str, Any]:
    args: dict[str, Any] = {
        "database_id": database_id,
        "page_size": limit or 100,
    }
    if filter:
        args["filter"] = filter
    if sort_property:
        args["sorts"] = [
            {"property": sort_property, "direction": sort_direction}
        ]
    return {"tool": "query_database", "args": args}

File: cli-creator/examples/node-search-cli/build-calls.ts
/**
 * Pure functions that transform CLI arguments into API call parameters.
 * No I/O, no side effects -- easy to test without mocking.
 */

export interface ApiCall {
  tool: string;
  args: Record<string, unknown>;
}

// --- Search ---

export interface SearchOptions {
  limit?: number;
  filter?: "page" | "database";
}

export function buildSearchCall(query: string, opts: SearchOptions = {}): ApiCall {
  return {
    tool: "search",
    args: {
      query,
      page_size: opts.limit ?? 20,
      ...(opts.filter && {
        filter: { property: "object", value: opts.filter },
      }),
    },
  };
}

// --- Fetch ---

export interface FetchOptions {
  includeChildren?: boolean;
}

export function buildFetchCall(id: string, opts: FetchOptions = {}): ApiCall {
  return {
    tool: "get_page",
    args: {
      page_id: id,
      ...(opts.includeChildren && { include_children: true }),
    },
  };
}

// --- Create ---

export interface CreateOptions {
  parent: string;
  title: string;
  body?: string;
}

export function buildCreateCall(opts: CreateOptions): ApiCall {
  return {
    tool: "create_page",
    args: {
      parent_id: opts.parent,
      properties: {
        title: opts.title,
      },
      ...(opts.body && {
        children: [
          {
            type: "paragraph",
            paragraph: {
              rich_text: [{ type: "text", text: { content: opts.body } }],
            },
          },
        ],
      }),
    },
  };
}

// --- Update ---

export interface UpdateOptions {
  id: string;
  title?: string;
  archived?: boolean;
}

export function buildUpdateCall(opts: UpdateOptions): ApiCall {
  const properties: Record<string, unknown> = {};
  if (opts.title !== undefined) {
    properties.title = opts.title;
  }

  return {
    tool: "update_page",
    args: {
      page_id: opts.id,
      ...(Object.keys(properties).length > 0 && { properties }),
      ...(opts.archived !== undefined && { archived: opts.archived }),
    },
  };
}

// --- Delete (Archive) ---

export function buildDeleteCall(id: string): ApiCall {
  return buildUpdateCall({ id, archived: true });
}

// --- Database Query ---

export interface QueryOptions {
  databaseId: string;
  filter?: Record<string, unknown>;
  sort?: { property: string; direction: "ascending" | "descending" };
  limit?: number;
}

export function buildQueryCall(opts: QueryOptions): ApiCall {
  return {
    tool: "query_database",
    args: {
      database_id: opts.databaseId,
      ...(opts.filter && { filter: opts.filter }),
      ...(opts.sort && {
        sorts: [
          { property: opts.sort.property, direction: opts.sort.direction },
        ],
      }),
      page_size: opts.limit ?? 100,
    },
  };
}
